File: arrays/stock.kt
fun main() {

    var prices = listOf(7, 1, 5, 3, 6, 4)
    println(maxProfit_nsquared(prices))
    println(maxProfit_linear(prices))
    println(maxProfit2(prices))

}


fun maxProfit_nsquared(prices: List<Int>): Int {
    var buy = 0
    var sell = 0
    var bestProfit = 0

    while (buy < prices.size) {
        while (sell < prices.size) {
            bestProfit = Math.max(prices[sell] - prices[buy], bestProfit)
            sell++
        }
        buy++
        sell = buy
    }
    
    return bestProfit
}


fun maxProfit_linear(prices: List<Int>): Int {
    var buy = 0
    var sell = 0
    var bestProfit = 0

    for (i in 1 until prices.size) {
        if (prices[i] <= prices[buy]) {
            buy = i
            sell = Math.max(buy, sell)
        }
        if (prices[i] >= prices[sell]) {
            sell = i
        }
        bestProfit = Math.max(prices[sell] - prices[buy], bestProfit)
    }
    
    return bestProfit
}


fun maxProfit2(prices: List<Int>): Int {
    var buy = 0
    var sell = 0
    var bestProfit = 0
    var maxProfit = 0

    for (i in 1 until prices.size) {
        if (prices[i] <= prices[buy]) {
            buy = i
            sell = Math.max(buy, sell)
        }
        if (prices[i] >= prices[sell]) {
            sell = i
        }
        if (bestProfit < prices[sell] - prices[buy]) {
            maxProfit += prices[sell] - prices[buy]
            buy = sell
            bestProfit = 0
        }
    }    
    return maxProfit
}

File: arrays/stock.py
#! /usr/bin/env python3

# Say you have an array for which the ith element is the price of a given
# stock on day i.
# If you were only permitted to complete at most one transaction
# (i.e., buy one and sell one share of the stock), design an algorithm to find
# the maximum profit.
# Note that you cannot sell a stock before you buy one.

class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        if not prices:
            return 0

        buy = 0
        max_profit = -2**31
        for i in range(1, len(prices)):
            cur_profit = prices[i] - prices[buy]
            if cur_profit < 0:
                buy = i
            else:
                max_profit = max(max_profit, cur_profit)

        return max_profit if max_profit > 0 else 0


if __name__ == "__main__":
    s = Solution()
    prices = [7, 2, 0, 1, 6]
    print(s.maxProfit(prices))
